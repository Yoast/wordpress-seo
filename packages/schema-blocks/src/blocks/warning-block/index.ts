import { WarningBlockAttributes } from "./warningBlockAttributes";

export { WarningBlockAttributes };
