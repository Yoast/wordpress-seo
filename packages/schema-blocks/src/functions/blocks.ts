import { select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}
