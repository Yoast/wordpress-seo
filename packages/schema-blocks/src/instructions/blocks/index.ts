import "./Block";
import "./ClassName";
import "./CurrencySelect";
import "./Date";
import "./Heading";
import "./InheritSidebar";
import "./InnerBlocks";
import "./RichText";
import "./Select";
import "./SidebarCheckbox";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./TextInput";
import "./Title";
import "./Variation";
import "./VariationPicker";
import "./Duration";
