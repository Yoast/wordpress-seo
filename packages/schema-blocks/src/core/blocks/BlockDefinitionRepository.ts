import BlockDefinition from "./BlockDefinition";

// Internal store of all known BlockDefinitions.
const registeredBlockDefinitions: Record<string, BlockDefinition> = {};

/**
 * Finds a registered BlockDefinition.
 *
 * @param name The Block to search for.
 *
 * @returns {BlockDefinition} The found block.
 */
export function getBlockDefinition( name: string ): BlockDefinition {
	return registeredBlockDefinitions[ name ];
}

/**
 * Registers a BlockDefinition.
 *
 * @param name       The name of the Definition to register.
 * @param definition The Definition to register.
 */
export function registerBlockDefinition( name: string, definition: BlockDefinition ): void {
	registeredBlockDefinitions[ name ] = definition;
}
