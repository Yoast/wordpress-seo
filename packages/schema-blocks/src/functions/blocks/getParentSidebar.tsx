import { BlockEditProps } from "@wordpress/blocks";
import { createElement } from "@wordpress/element";
import { ReactElement } from "react";
import { getBlockDefinition } from "../../core/blocks/BlockDefinitionRepository";
import { getBlockByClientId } from "../BlockHelper";
import { createBlockEditProps, getParentIdOfType } from "../gutenberg/block";
import logger from "../logger";
import { PanelBody } from "@wordpress/components";

/**
 * Gets the sidebar from the parent of a blocks.
 *
 * @param props The block props.
 * @param parents The parents to get the sidebar from.
 *
 * @returns The sidebar.
 */
export default function getParentSidebar( props: BlockEditProps<unknown>, parents: string[] ): ReactElement {
	let parentIds: string[] = [];
	if ( parents ) {
		parentIds = getParentIdOfType( props.clientId, parents );
	}

	const elements: ReactElement[] = [];
	if ( parentIds.length > 0 ) {
		parentIds.forEach( parentId => {
			const parentBlock = getBlockByClientId( parentId );
			const parentBlockDefinition = getBlockDefinition( parentBlock.name );
			if ( parentBlockDefinition ) {
				logger.debug( props.clientId + " inherited sidebar from " + parentBlock.name + " definition" );
				const parentProps = createBlockEditProps( parentBlock );
				elements.push( ...parentBlockDefinition.sidebarElements( parentProps ) );
			}
		} );
	}

	return <PanelBody>{ ...elements }</PanelBody>;
}
