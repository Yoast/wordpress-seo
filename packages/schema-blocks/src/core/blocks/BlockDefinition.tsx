import { ReactElement } from "react";
import { createElement, Fragment } from "@wordpress/element";
import {
	registerBlockType,
	BlockConfiguration,
	BlockEditProps,
	BlockSaveProps,
} from "@wordpress/blocks";
import { InspectorControls, BlockIcon } from "@wordpress/block-editor";
import { select } from "@wordpress/data";
import { PanelBody } from "@wordpress/components";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";
import { registerBlockDefinition } from "./BlockDefinitionRepository";
import logger from "../../functions/logger";
import { openGeneralSidebar } from "../../functions/gutenberg/sidebar";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId: string;
	name?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
	className?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class.
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "b", "c", "d", "f", "g", "h", "k", "m", "z" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): ReactElement {
		// Force the sidebar open.
		if ( select( "core/block-editor" ).isBlockSelected( props.clientId ) ) {
			openGeneralSidebar( "edit-post/block", true );
		}

		const sidebarElements = this.sidebarElements( props );

		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		if ( sidebarElements.length > 0 ) {
			const sidebarContainer =
				<InspectorControls key="sidebar-inspector-controls">
					<PanelBody>
						{ sidebarElements }
					</PanelBody>
				</InspectorControls>;

			elements.unshift( sidebarContainer );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as ReactElement;
		}

		return createElement( Fragment, { key: props.clientId }, elements.map( ( element, i ) => createElement( Fragment, { key: i }, element ) ) );
	}

	/**
	 * Renders the persisted block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): ReactElement {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;
		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		logger.debug( "registering block " + name );

		if ( configuration.icon && typeof configuration.icon === "string" && configuration.icon.startsWith( "<svg" ) ) {
			configuration.icon = this.createBlockIcon( configuration );
		}

		// Register the block to WordPress.
		registerBlockType( name, configuration );
		// Register the block with our own code.
		registerBlockDefinition( name, this );
	}

	/**
	 * Creates the sidebar elements.
	 *
	 * @param props The properties of the block to create a sidebar for.
	 *
	 * @returns The sidebar element to render.
	 */
	sidebarElements( props: RenderEditProps ): ReactElement[] {
		return Object.values( this.instructions )
			.map( ( instruction, index ) => <Fragment key={ instruction.id }>
				{ instruction.sidebar( props, index ) }
			</Fragment> )
			.filter( e => e !== null );
	}

	/**
	 * Creates a block icon.
	 *
	 * @param configuration The block configuration.
	 *
	 * @returns The sidebar element to render.
	 */
	private createBlockIcon( configuration: MutableBlockConfiguration ): ReactElement {
		const icon = <span
			className="yoast-schema-blocks-icon"
			dangerouslySetInnerHTML={ { __html: configuration.icon as string } }
		/>;

		return <BlockIcon icon={ icon } />;
	}
}
