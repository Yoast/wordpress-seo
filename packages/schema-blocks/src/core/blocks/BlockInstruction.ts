import BlockLeaf from "./BlockLeaf";
import { RenderEditProps, RenderSaveProps } from "./BlockDefinition";
import { ReactElement } from "react";
import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";
import { BlockValidation, BlockValidationResult, BlockPresence } from "../validation";
import Instruction from "../Instruction";

/**
 * BlockInstruction class.
 */
export default abstract class BlockInstruction extends Instruction {
	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders saving the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	save( props: RenderSaveProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders editing the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	edit( props: RenderEditProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders the sidebar.
	 *
	 * @param props The props.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The sidebar element to render.
	 */
	sidebar( props: RenderEditProps, i: number ): ReactElement {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/**
	 * Returns the configuration of this instruction.
	 *
	 * @returns The block configuration.
	 */
	configuration(): Partial<BlockConfiguration> {
		return {
			attributes: {
				[ this.options.name ]: {
					required: this.options.required === true,
				},
			},
		};
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns {BlockValidationResult} The validation result.
	 */
	validate( blockInstance: BlockInstance ): BlockValidationResult {
		return new BlockValidationResult( blockInstance.clientId, this.constructor.name, BlockValidation.Unknown, BlockPresence.Unknown );
	}
}
