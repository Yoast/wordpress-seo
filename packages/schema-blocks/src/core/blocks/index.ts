import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
};
