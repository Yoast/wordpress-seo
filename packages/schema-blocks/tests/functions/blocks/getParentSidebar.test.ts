import "../../matchMedia.mock";
import { BlockEditProps, BlockInstance } from "@wordpress/blocks";
import * as renderer from "react-test-renderer";
import getParentSidebar from "../../../src/functions/blocks/getParentSidebar";

jest.mock( "../../../src/functions/gutenberg/block", () => {
	return {
		getParentIdOfType: jest.fn( ( clientId )  => {
			if ( clientId === "404" ) {
				return [];
			}

			return [ clientId ];
		} ),
		createBlockEditProps: jest.fn( ( block, selected = false )  => {
			return {
				clientId: block.clientId,
				isSelected: selected,
			} as BlockEditProps<unknown>;
		} ),
	};
} );

jest.mock( "../../../src/functions/BlockHelper", () => {
	return {
		getBlockByClientId: jest.fn( ( clientId )  => {
			if ( clientId === "1337" ) {
				return {
					clientId: clientId,
					name: "yoast/block",
				} as BlockInstance;
			}

			return {
				clientId: clientId,
				name: "yoast/unknown-block",
			} as BlockInstance;
		} ),
	};
} );

jest.mock( "../../../src/core/blocks/BlockDefinitionRepository", () => {
	return {
		getBlockDefinition: jest.fn( ( blockName )  => {
			if ( blockName === "yoast/block" ) {
				return {
					clientId: "1337",
					sidebarElements: jest.fn( ( props: BlockEditProps<unknown> ) => {
						return [ props.clientId ];
					} ),
				};
			}

			return null;
		} ),
	};
} );

describe( "The getParentSidebar function", () => {
	it( "receives the parents argument as null", () => {
		const props: BlockEditProps<unknown> = {
			className: "",
			clientId: "1337",
			isSelected: false,
			setAttributes: null,
			attributes: null,
		};

		const tree = renderer
			.create( getParentSidebar( props, null ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );
	it(  "doesn't find parentIds for given parents", () => {
		const props: BlockEditProps<unknown> = {
			className: "",
			clientId: "404",
			isSelected: false,
			setAttributes: null,
			attributes: null,
		};

		const tree = renderer
			.create( getParentSidebar( props, [ "yoast/block" ] ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );
	it(  "receives finds a parent without any block definition", () => {
		const props: BlockEditProps<unknown> = {
			className: "",
			clientId: "107",
			isSelected: false,
			setAttributes: null,
			attributes: null,
		};

		const tree = renderer
			.create( getParentSidebar( props, [ "yoast/block" ] ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );
	it(  "receives finds a parent and retrieves its sidebar", () => {
		const props: BlockEditProps<unknown> = {
			className: "",
			clientId: "1337",
			isSelected: false,
			setAttributes: null,
			attributes: null,
		};

		const tree = renderer
			.create( getParentSidebar( props, [ "yoast/block" ] ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );
} );
