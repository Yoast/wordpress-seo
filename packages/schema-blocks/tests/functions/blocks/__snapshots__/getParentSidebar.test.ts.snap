// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The getParentSidebar function doesn't find parentIds for given parents 1`] = `
<div
  className="components-panel__body is-opened"
/>
`;

exports[`The getParentSidebar function receives finds a parent and retrieves its sidebar 1`] = `
<div
  className="components-panel__body is-opened"
>
  1337
</div>
`;

exports[`The getParentSidebar function receives finds a parent without any block definition 1`] = `
<div
  className="components-panel__body is-opened"
/>
`;

exports[`The getParentSidebar function receives the parents argument as null 1`] = `
<div
  className="components-panel__body is-opened"
/>
`;
